File: cp35/cses/Counting_Rooms.cpp
#include <bits/stdc++.h>
using namespace std;
#define watch(x) cout << (#x) << " is " << (x) << "\n"
#define watch2(x, y) cout << #x << "=" << x << "," << #y << "=" << y << "\n"


// 5 8
// ########
// #..#...#
// ####.#.#
// #..#...#
// ########
// represent height and width of the room
int n, m, rooms;
vector<vector<bool>> visited; // tracks visited and impossible to visit
vector<pair<int, int>> moves = {
    {-1, 0},  // up
    {1, 0 },  // down
    {0, -1},  // left
    {0, 1}    //  right
  };


bool isValid(int x, int y){
  // is out of bounds
  if(x < 0 || x >= n || y < 0 || y >= m){
    return false;
  }
  // it has been visited
  if(visited[x][y]){
    return false;
  }
  // neither vistied nor out of bounds
  return true;
}

void dfs(int row, int col){
  visited[row][col] = true;
  for(auto move: moves){
    if(isValid(row + move.first , col + move.second)){
      dfs(row + move.first, col + move.second);
    }
  }
}


void solve(){
  for(int i = 0; i < n; i++){
    for(int j = 0; j < m; j++){
      if(!visited[i][j]){
        // for each location map the room
        dfs(i, j);
        rooms++;
      }
    }
  }
}



int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cin >> n >> m;
    visited.resize(n);

    for(int i = 0; i < n; i++){
      visited[i].resize(m);
    }
 


    for(int i = 0; i < n; i++){
      for(int j = 0; j < m; j++){
        char c; 
        cin >> c;
        if(c == '#'){
          visited[i][j] = 1;
        }
      }
    }
    solve();
    cout << rooms << "\n";
}

File: cp35/python/graphs/counting_rooms.py



moves = [
    (-1, 0),  # up
    (1, 0 ),  # down
    (0, -1),  # left
    (0, 1)    #  right
  ]

def is_valid(x,y):
  if x < 0 or x >=n  or y < 0 or y >= m:
    return False
  if visited[x][y] == True:
    return False
  return True

def dfs(x,y):
  visited[x][y] = True
  for move in moves:
    row = x + move[0]
    col = y + move[1]
    if is_valid(row, col):
      dfs(row, col)
    

n = 0
m = 0


def count_rooms():
  count = 0
  for i in range(0, n):
    for j in range(0, m):
      if visited[i][j] == False:
        dfs(i,j)
        count += 1
  return count






visited = []
if __name__ == '__main__':
    n, m = map(int, input().split())
    for i in range(n):
        arr = list(input())
        visited.append(arr)
    
    for i in range(n):
      for j in range(m):
        if visited[i][j] == '.':
          visited[i][j] = False
        else:
          visited[i][j] = True
    print(count_rooms())
    


